pokedex/storage.lua

```diff
@@ -8,11 +8,17 @@ local log = require "utils.log"
 
 local M = {}
 
+-- All of these tables are saved out to a file, so be sure they are raw data only (no functions, userdata, that sort of thing)
+local storage_data = {} -- contains all the below tables
 local player_pokemon = {}
+local pokemon_by_location = {}
```

Comment:
You currently have these set up as `pc` and `party` arrays inside the `pokemon_by_location` table; however, I think an actual table would be more effective. The reason I say this is because there are several places where you loop through all the pokemon in a location to find just one. That is, again, slow when searching through a PC full of 1000 pokemon. If you are going to store it in a list like that, it either should be:
* sorted (like by id) so you can find the pokemon quickly, or
* a table (indexed by id)

I recommend the table version for this case, as there's no real reason this needs to be an ordered list (right?). In that case you might want to store like so:

```
-- when adding
pokemon_by_location.pc[id] = true

-- when removing
pokemon_by_location.pc[id] = nil
```

Reply:
I was thinking of doing `pokemon_by_location.pc[id]` too first but I didn't end up doing I know that it's _incredibly_ slow when looping.

You comment made me want to confirm my suspicion so I did some quick profiling of it and `for id, _ in pairs(...) do` is about _60_ times slower than `for i=1, #... do`. Of course finding a id is a lost faster with a look up (about 120 times faster) but finding the id is still very fast (0.007 vs 0.86, while loop was 53.12 vs 0.84).

I threw the test into this gist so you can take a look (and make sure I don't do something odd) https://gist.github.com/Jerakin/5764997c9de6936c7e4a5e6db721b511

Here is the full output of the script
```lua
DEBUG:SCRIPT: find_by_id: first      -- first id in table[id]
DEBUG:SCRIPT: 0.0070000000391701
DEBUG:SCRIPT: find_by_id: mid        -- mid id in table[id]
DEBUG:SCRIPT: 0.0019999999949505
DEBUG:SCRIPT: find_by_id: last       -- last id in table[id]
DEBUG:SCRIPT: 0.00099999999747524
DEBUG:SCRIPT: find_by_index: first   -- first id in table[n]
DEBUG:SCRIPT: 0.0019999999949505
DEBUG:SCRIPT: find_by_index: mid     -- mid id in table[n]
DEBUG:SCRIPT: 0.86199999998371
DEBUG:SCRIPT: find_by_index: last    -- last id in table[n]
DEBUG:SCRIPT: 1.4590000000112
DEBUG:SCRIPT: loop: loop_all_id      -- for id, _ in pairs(by_id) do
DEBUG:SCRIPT: 53.138999999987
DEBUG:SCRIPT: loop: loop_all_index   -- for i=1, #by_index do
DEBUG:SCRIPT: 0.84499999996979
```

Reply:
Cool, I'll put some comments in that other review.

Reply:
Changed it to use the "hash table" as you proposed.